File: gps_agent_pkg/src/pytorchcontroller.cpp
#include "gps_agent_pkg/pytorchcontroller.h"
#include "gps_agent_pkg/robotplugin.h"
#include "gps_agent_pkg/util.h"

#include <ros/ros.h>
#include <torch/version.h>

using namespace gps_control;

// ---------------------------------------------------------------------------
// Constructor / destructor
// ---------------------------------------------------------------------------

PyTorchController::PyTorchController()
    : TrialController()
{
    is_configured_ = false;
}

PyTorchController::~PyTorchController() = default;

// ---------------------------------------------------------------------------
// configure_controller
// ---------------------------------------------------------------------------

void PyTorchController::configure_controller(OptionsMap &options)
{
    // Let the base class handle T, state/obs datatypes, ee_tgt, etc.
    TrialController::configure_controller(options);

    // ---- Validate Python/C++ torch version compatibility ---------------
    const std::string python_torch_version =
        std::get<std::string>(options.at("torch_version"));
    int py_major = 0, py_minor = 0;
    std::sscanf(python_torch_version.c_str(), "%d.%d", &py_major, &py_minor);
    if (py_major != TORCH_VERSION_MAJOR) {
        ROS_ERROR(
            "PyTorchController: TorchScript major version mismatch: "
            "Python torch %d.%d vs C++ LibTorch %d.%d — refusing to load model",
            py_major, py_minor, TORCH_VERSION_MAJOR, TORCH_VERSION_MINOR);
        is_configured_ = false;
        return;
    }
    if (py_minor != TORCH_VERSION_MINOR) {
        ROS_WARN(
            "PyTorchController: TorchScript minor version mismatch: "
            "Python torch %d.%d vs C++ LibTorch %d.%d — attempting to load anyway",
            py_major, py_minor, TORCH_VERSION_MAJOR, TORCH_VERSION_MINOR);
    }

    // ---- Load TorchScript model from raw bytes -------------------------
    const std::string model_bytes =
        std::get<std::string>(options.at("model_bytes"));

    try {
        std::istringstream stream(model_bytes);
        module_ = torch::jit::load(stream, torch::kCPU);
        module_.eval();
        module_loaded_ = true;
    } catch (const std::exception &e) {
        ROS_ERROR("PyTorchController: failed to load TorchScript model: %s",
                  e.what());
        module_loaded_ = false;
        return;
    }

    // ---- Observation normalisation ------------------------------------
    // scale is transmitted as the diagonal of the scale matrix (length dO).
    scale_diag_ = std::get<Eigen::VectorXd>(options.at("scale"));
    bias_        = std::get<Eigen::VectorXd>(options.at("bias"));

    // ---- Per-timestep noise ------------------------------------------
    int T = std::get<int>(options.at("T"));
    noise_.resize(T);
    for (int t = 0; t < T; ++t) {
        noise_[t] = std::get<Eigen::VectorXd>(options.at("noise_" + std::to_string(t)));
    }

    // Infer dU from the first noise vector.
    dU_ = (T > 0) ? static_cast<int>(noise_[0].size()) : 0;

    ROS_INFO_STREAM("PyTorchController: loaded TorchScript model, dU=" << dU_
                    << ", T=" << T);
    is_configured_ = true;
}

// ---------------------------------------------------------------------------
// get_action
// ---------------------------------------------------------------------------

void PyTorchController::get_action(int t,
                                   const Eigen::VectorXd & /*X*/,
                                   const Eigen::VectorXd &obs,
                                   Eigen::VectorXd &U)
{
    if (!is_configured_ || !module_loaded_) {
        // Safety: output zero torques until we are properly configured.
        if (dU_ > 0) U.setZero(dU_);
        return;
    }

    const int dO = static_cast<int>(obs.size());

    // ---- Apply observation normalisation ----
    // obs_scaled = obs .* scale_diag + bias
    // scale_diag_ stores the diagonal of the scale matrix (pointwise multiply).
    Eigen::VectorXd obs_scaled(dO);
    for (int i = 0; i < dO; ++i) {
        obs_scaled(i) = obs(i) * scale_diag_(i) + bias_(i);
    }

    // ---- Build input tensor [1, dO] on CPU ----
    auto options_t = torch::TensorOptions().dtype(torch::kFloat32);
    torch::Tensor input = torch::zeros({1, dO}, options_t);
    {
        auto acc = input.accessor<float, 2>();
        for (int i = 0; i < dO; ++i) {
            acc[0][i] = static_cast<float>(obs_scaled(i));
        }
    }

    // ---- Forward pass (no_grad, CPU, eval mode) ----
    torch::Tensor output;
    try {
        torch::NoGradGuard no_grad;
        output = module_.forward({input}).toTensor();
    } catch (const std::exception &e) {
        ROS_ERROR_THROTTLE(1.0, "PyTorchController: forward pass failed: %s",
                           e.what());
        U.setZero(dU_);
        return;
    }

    // ---- Unpack output [1, dU] → U ----
    U.resize(dU_);
    {
        auto acc = output.accessor<float, 2>();
        for (int i = 0; i < dU_; ++i) {
            U(i) = static_cast<double>(acc[0][i]);
        }
    }

    // ---- Add pre-sampled noise ----
    if (t >= 0 && t < static_cast<int>(noise_.size())) {
        U += noise_[t];
    }
}

File: python/gps_agent_pkg/pytorch_controller.py
"""
Translates gps_agent_pkg/src/pytorchcontroller.cpp + include/pytorchcontroller.h.

Key logic:
  1. configure_controller: load TorchScript model from raw bytes
  2. Version check: torch.__version__ major must match the torch_version field
  3. get_action(t, X, obs):
       obs_scaled = obs * scale_diag + bias    (pointwise affine normalisation)
       U = module(obs_scaled)[0] + noise_[t]
"""
from __future__ import annotations

import io
import logging
import time

import numpy as np

from gps_agent_pkg.trial_controller import TrialController

LOGGER = logging.getLogger(__name__)

try:
    import torch
    _TORCH_AVAILABLE = True
except ImportError:
    _TORCH_AVAILABLE = False


class PyTorchController(TrialController):
    """
    TorchScript-policy trial controller.

    Mirrors C++ gps_control::PyTorchController.
    """

    MAX_INFERENCE_MS: float = 5.0

    def __init__(self) -> None:
        super().__init__()
        self._module = None
        self._scale_diag: np.ndarray = np.array([], dtype=np.float64)
        self._bias:       np.ndarray = np.array([], dtype=np.float64)
        self._noise:      list[np.ndarray] = []
        self._dU: int = 0

    # ------------------------------------------------------------------
    # Configuration
    # ------------------------------------------------------------------

    def configure_controller(self, options: dict) -> None:
        """
        Load TorchScript model and read normalisation / noise parameters.

        Expected keys in *options*:
          "model_bytes"   : bytes | str  — raw TorchScript bytes
          "torch_version" : str          — e.g. "2.1.0"
          "scale"         : array-like   — diagonal of scale matrix (length dO)
          "bias"          : array-like   — bias vector (length dO)
          "T"             : int          — trial length
          "noise_{t}"     : array-like   — noise for timestep t, t=0..T-1
        """
        super().configure_controller(options)

        if not _TORCH_AVAILABLE:
            raise ImportError("torch is required for PyTorchController")

        # ---- Version check -------------------------------------------
        torch_version: str = str(options["torch_version"])
        model_major = int(torch_version.split(".")[0])
        runtime_major = int(torch.__version__.split(".")[0])
        if model_major != runtime_major:
            raise ValueError(
                f"Torch major version mismatch: model={model_major} "
                f"runtime={runtime_major}"
            )

        # ---- Load model from raw bytes --------------------------------
        raw = options["model_bytes"]
        if isinstance(raw, (bytes, bytearray)):
            buf = io.BytesIO(raw)
        else:
            # Handle list[int] (how ROS messages carry byte arrays)
            buf = io.BytesIO(bytes(raw))
        buf.seek(0)
        self._module = torch.jit.load(buf, map_location="cpu")
        self._module.eval()

        # ---- Normalisation -------------------------------------------
        self._scale_diag = np.asarray(options["scale"], dtype=np.float64)
        self._bias       = np.asarray(options["bias"],  dtype=np.float64)

        # ---- Per-timestep noise --------------------------------------
        T = int(options["T"])
        self._noise = [
            np.asarray(options[f"noise_{t}"], dtype=np.float64) for t in range(T)
        ]
        self._dU = len(self._noise[0]) if T > 0 else 0

        self.is_configured_ = True

    # ------------------------------------------------------------------
    # Action
    # ------------------------------------------------------------------

    def get_action(self, t: int, X: np.ndarray, obs: np.ndarray) -> np.ndarray:
        """
        Normalise obs, run the TorchScript model, add noise.

        Mirrors C++ PyTorchController::get_action.
        """
        if not self.is_configured_ or self._module is None:
            return np.zeros(self._dU, dtype=np.float64)

        # 1. Observation normalisation: obs_scaled = obs * diag(scale) + bias
        obs_scaled = obs * self._scale_diag + self._bias

        # 2. Build [1, dO] float32 input tensor
        inp = torch.tensor(obs_scaled, dtype=torch.float32).unsqueeze(0)

        # 3. Forward pass with timing
        t0 = time.perf_counter()
        with torch.no_grad():
            out = self._module(inp)
        elapsed_ms = (time.perf_counter() - t0) * 1000.0
        if elapsed_ms > self.MAX_INFERENCE_MS:
            LOGGER.warning(
                "PyTorchController: inference %.2f ms > %.1f ms threshold",
                elapsed_ms, self.MAX_INFERENCE_MS,
            )

        # 4. Unpack [1, dU] → (dU,) float64
        U = out.squeeze(0).numpy().astype(np.float64)

        # 5. Add pre-sampled noise
        if 0 <= t < len(self._noise):
            U = U + self._noise[t]

        return U
